### DISCUSSION TITLE
First candidate layout of 24 home block letters

### DISCUSSION THREAD

**Discussion by binarybottle** [binarybottle/engram-es#9] [Q&A]
I just generated a first candidate layout of 24 home block letters for Spanish.  For native speakers (alert to @NickG13, @Lobo-Feroz, and @sunaku) please tell me if there are any glaring concerns, or if there is anything awkward or offensive that happens to have arisen in the layout of letters:

        H Y O Z  M L B J
        P I E A  D S N R
        G Q X U  T C V F
        
          7   7  74   3   22  48  10   4
         23  57 109 103   42  61  58  55
         10   7   2  33   36  36   8   7
        
        left: 43336111 (11.72%)  right: 38788861
        Total same-finger bigram frequencies:    106090282428
        Total bigram inward roll frequencies:   6394006706059

**Comment by Lobo-Feroz** [comment]
Looks awesome! Let's give it a try!

thanks for everything @binarybottle !

**Comment by binarybottle** [comment]
W above K received the highest score:

Rank                                                                   Score
1:  H Y O Z P I E A G Q X U M L B J D S N R T C V F ' , - " . ? W K    0.6315284924483411
1:  J B L M R N S D F V C T Z O Y H A E I P U X Q G ' , - " . ? W K    0.6315284472608467
1:  H Y O Z P I E A G Q X U M L B J D S N R T C V F ' , - " . ? K W    0.6315270680131808
1:  J B L M R N S D F V C T Z O Y H A E I P U X Q G ' , - " . ? K W    0.631526057420239

So the candidate letters of the Engram-es layout are:

    H Y O Z  M L B J W
    P I E A  D S N R K
    G Q X U  T C V F

**Reply by Lobo-Feroz** [reply]
First draft of a windows layout. Attached .klc file.

The punctuation is almost the same as engram-en with some changes that we commented (just for testing the keyboard).

![engram-es-v1-normal](https://user-images.githubusercontent.com/10242383/127710372-b1f838d6-525f-4418-aa95-3deb6b318714.png)

![imagen](https://user-images.githubusercontent.com/10242383/127710053-29fae3af-c2f3-4cb1-a049-8b0675770da8.png)

[engram-es-v1-20200730.zip](https://github.com/binarybottle/engram-es/files/6910028/engram-es-v1-20200730.zip)

**Reply by binarybottle** [reply]
Thank you very much!  Don't get too attached to this initial candidate, As I am crunching through over 40 million more layouts over the course of this week based on this one, and we will find other candidates to compare against.

**Reply by NickG13** [reply]
This looks awesome! 
Just to test it and also practice and learn this layout I've encountered a huge tool, it's called **Type Fu**.

I will add here two practice files for spanish words, that you can add into the app itself (on custom words section)
[Type-Fu](https://type-fu.com/)
I was using this to test and train AEIOU layout, we could also use it in the long therm for the definitive one, but it's awesome to gain some speed and see how the layout performs after finger travel comfort. 
[Spanish training.zip](https://github.com/binarybottle/engram-es/files/6910754/Spanish.training.zip) we will only need the layout file. I will surely create this in a few hours and upload. 
UPDATE: Here is the layout file
[Engram-es.zip](https://github.com/binarybottle/engram-es/files/6910781/Engram-es.zip)
It shall look something like this 
![image](https://user-images.githubusercontent.com/12291843/127730952-c91fc45e-c33b-4832-aae0-22538fbfe19a.png)

**Reply by Lobo-Feroz** [reply]
@NickG13 I've tried your layout and word files and they work nicely. And I have the initial, subjective feeling that this layout will be very easy on the hands.

Also, I propose that we move the typing learning discussion to its own thread.

**Reply by ficion** [reply]
Would the bigram 'qu' be comfortable to type in this layout?

**Reply by binarybottle** [reply]
I am waiting for a refined set of bigram frequencies, and will replace this tentative layout with a better one...

**Comment by binarybottle** [comment]
Please see https://github.com/binarybottle/engram-es/discussions/7#discussioncomment-1100537 for suggested symbols and diacriticals.

**Comment by binarybottle** [comment]
@NickG13 @Lobo-Feroz @iandoug --

I am running a stability test to see if I can find higher-scoring layouts, but here is the preliminary optimal Spanish layout based on the cleaned-up text corpus:

      { | = ~ +   <  >   ^ & % * } \
      [ 1 2 3 4   5  6   7 8 9 0 ] /

        Z H O X  '(  ")  M D B J W #€$ @`
        P I E A  ,;  .:  T S N R K
        F Y Q U  -_  ?!  C L V G

Next: finalize non-letter characters!  

What is the current stance on diacritical trigger key(s)?

**Reply by iandoug** [reply]
Looks reasonable, don't see anything concerning. :-)

Oh wait, a lot of words and in A, so 'A,' probably going to trigger high SFB.  ?

Consider D on home row rather than T.

Cheers, Ian

**Reply by binarybottle** [reply]
@NickG13 @Lobo-Feroz -- Do you see anything offensive, slang, abbreviations, or otherwise in this layout?

**Reply by binarybottle** [reply]
@iandoug -- What is SFB?

**Reply by iandoug** [reply]
Same-Finger bigrams

Used a lot on https://www.reddit.com/r/KeyboardLayouts

**Reply by binarybottle** [reply]
ST is a much more frequent bigram than SD or DS, so it makes sense to avoid ST as a same-finger bigram.

**Reply by NickG13** [reply]
I'm late to the party! I would consider maybe trying to put C in home row also. I've seen it's very frequent with R and other vocals.  
As per UA/AU (not even considering AX or XA) or other combination with near keys with index finger, as I see with ian stadistics it will have a higher SFB with comma and period than E 

I have not seen any slang or offensive abbreviation on this layout, neither with PIAE if you finally change it.

**Reply by Lobo-Feroz** [reply]
Awesome to have a first version! 

> Do you see anything offensive, slang, abbreviations, or otherwise in this layout?

Nothing with any meaning pops up.

**Comment by binarybottle** [comment]
I'm not as concerned about same-finger bigrams with an intervening space as I am with the bigram with the preceding letter, as I believe that the cadence is quite different.
________________________________

**Reply by iandoug** [reply]
? there is no space between a and comma .

_Don Serafín Estébanez Calderón, político y conspirador, novelista,
historiador y poeta, nació en Málaga en 1799, y murió en Madrid en 1867.

**Reply by binarybottle** [reply]
What that tells me is that we should place the comma on the opposite side of the cluster of vowels or consonants. Does the comma follow consonants or vowels more often?

**Reply by iandoug** [reply]
Counts
Period
```
s.	4939161
o.	4638801
a.	4571062
e.	1835866
n.	1282300
r.	1139892
l.	889404
d.	565374
.-	461045
".	390942
”.	266059
ó.	246577
D.	220996
z.	185657
._	164650
_.	158264
```
Comma
```
s,	7683043
o,	7663386
a,	7002247
e,	3241682
n,	1947843
r,	1515896
l,	1178556
d,	712561
",	447128
í,	376064
y,	375873
z,	370399
_,	328883
”,	314995
),	236985
ó,	204211
i,	182035
```

**Reply by binarybottle** [reply]
Thank you, @iandoug -- this is exactly what I needed!

**Reply by binarybottle** [reply]
For o, a, and e, it looks like there are a lot more following commas than periods, so I should switch the position of the comma and period in the layout.

**Reply by NickG13** [reply]
Of course you will also find always a higher percentage of commas on a text, thats logical too. I would change A with E instead, but do what you consider of course. I also think period would be better in left (indeed, thats how I designed my layout)

**Reply by NickG13** [reply]
I would also joint both " frequencies as the same character, since in spanish we use it as the same thing

**Reply by iandoug** [reply]
> I would also joint both " frequencies as the same character, since in spanish we use it as the same thing

They are separate now as requested... :-)

**Reply by iandoug** [reply]
> Thank you, @iandoug -- this is exactly what I needed!

That was the point of my original bigram file.

Also so that you could see which capitals are common, and put them on appropriate hand ... distance to Shift is not the same Left/Right ANSI/ISO ...

cheers, Ian

**Reply by Lobo-Feroz** [reply]
I think some degree of SFB will be unavoidable.

As per Ian's numbers we see that most words end in o, a, e. This is because theres letters denote masculine, femenine, neutral. "est**a**, est**o**, est**e**" (this, this, this). And the most ending is in 's' because it's the plural (esto**s** - these). And 'r' shows up a lot because in spanish many verbs end in 'r'. 

So the fact of the matter is that the most frequent letters are also the most frequent word ending letters, so I don't think this is avoidable except by not having popular letters in the home row index finger position (which doesn't make much sense), or moving period and comma elsewhere, which doesn't make much sense either, since center columns home row seem a good placement for the most used punctuation signs.

In an ergo keyboard maybe we could move them to the thumb or something, with spaces, the three of them going after the ending letter of a word. But for ISO/ANSI I don't know which could be the best compromise. 

Another option could be including the main punctuation signs in the optimization algorithm, but that would yield a keyboard where the signs could be anywhere, different from the engram signature of having them in the center columns.

**Reply by binarybottle** [reply]
Thank you all.  I'm looking at the punctuation/diacritical frequencies now to guide placement. The frequencies above suggest that the comma should probably go on the right.

**Comment by Lobo-Feroz** [comment]
Made a test layout for type-fu.com with the newest @binarybottle 's layout, and added the comma-period key and the accute accent key to test it. 

Here's the type fu layout:

[Engram-ES - V0 - 20200820-03.tfl.zip](https://github.com/binarybottle/engram-es/files/7023906/Engram-ES.-.V0.-.20200820-03.tfl.zip)

![engram-es typefu V0 20200820](https://user-images.githubusercontent.com/10242383/130288352-ac171f1a-0fc1-4a2e-b240-a66c79c3124f.png)

I've tested it with @NickG13 's spanish dictionary and so far seems easy on the fingers. The home row is very easy to memorize.

**Reply by binarybottle** [reply]
That's great news! Next up is to assign non-letter characters to the remaining half dozen or so keys!

**Reply by iandoug** [reply]
is this using "dead key" approach for the diacritics?

**Reply by Lobo-Feroz** [reply]
@binarybottle yeah, I didn't want to get ahead of myself, since having comma, period and the diacritics in the home row is just a proposal for the time being.

I wanted to know what did you guys think, and if there's some kind of criterion on which of these keys and its layers are easier-harder.

@iandoug yeah, i just placed all the diacritics, as dead key, in the home row center column, right index finger, since all the vowels are on the left. base: accute accent. shifted: diaeresis. altgr: grave accent. algr+shift: circumflex. 

This is just an idea of course, because spanish does not use the last two ones, we could do something else like the altgr of that key being the apostrophe (but I think one of you proposed the apostrope ' as the altgr of the quotes ")

**Reply by iandoug** [reply]
I will need to modify KLA to handle dead keys ... it has no concept of them. Will also need .json and UI changes to add them.

**Reply by Lobo-Feroz** [reply]
> I will need to modify KLA to handle dead keys ... it has no concept of them. Will also need .json and UI changes to add them.

I think that if you want to have international support, this would be the way to go. Even if all new keyboards used a different approach for diacriticals, you would need to support all the existing qwerty-es, -fr, -de etc. even if only for comparison purposes.

I don't know if I could help you with that. I imagine that must be more of a legacy code base issue.

**Reply by NickG13** [reply]
I don't know any analyzer that has support for dead keys, so that would be surely a milestone in this category

**Reply by iandoug** [reply]
Adding support is non-trivial... starting with the UI. :-)

**Reply by NickG13** [reply]
Yeah, that makes it even more a milestone!
